runtime/compiler/runtime/JITServerAOTDeserializer.cpp

```diff
@@ -898,3 +901,593 @@ JITServerLocalSCCAOTDeserializer::updateSCCOffsets(SerializedAOTMethod *method,
 
    return true;
    }
+
+JITServerNoSCCAOTDeserializer::JITServerNoSCCAOTDeserializer(TR_PersistentClassLoaderTable *loaderTable) :
+   JITServerAOTDeserializer(loaderTable),
+   _classLoaderIdMap(decltype(_classLoaderIdMap)::allocator_type(TR::Compiler->persistentAllocator())),
+   _classLoaderPtrMap(decltype(_classLoaderPtrMap)::allocator_type(TR::Compiler->persistentAllocator())),
+   _classIdMap(decltype(_classIdMap)::allocator_type(TR::Compiler->persistentAllocator())),
+   _classPtrMap(decltype(_classPtrMap)::allocator_type(TR::Compiler->persistentAllocator())),
+   _methodIdMap(decltype(_methodIdMap)::allocator_type(TR::Compiler->persistentAllocator())),
+   _methodPtrMap(decltype(_methodPtrMap)::allocator_type(TR::Compiler->persistentAllocator())),
+   _classChainMap(decltype(_classChainMap)::allocator_type(TR::Compiler->persistentAllocator())),
+   _wellKnownClassesMap(decltype(_wellKnownClassesMap)::allocator_type(TR::Compiler->persistentAllocator()))
+   { }
+
+
+void
+JITServerNoSCCAOTDeserializer::invalidateClassLoader(J9VMThread *vmThread, J9ClassLoader *loader)
+   {
+   assertExclusiveVmAccess(vmThread);
+
+   auto p_it = _classLoaderPtrMap.find(loader);
+   if (p_it == _classLoaderPtrMap.end())// Not cached
+      return;
+
+   uintptr_t id = p_it->second;
+   auto i_it = _classLoaderIdMap.find(id);
+   TR_ASSERT(i_it != _classLoaderIdMap.end(), "Broken two-way map");
+
+   // Mark entry as unloaded
+   i_it->second = NULL;
+   _classLoaderPtrMap.erase(p_it);
+
+   if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+      TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "Invalidated class loader %p ID %zu", loader, id);
+   }
+
+void
+JITServerNoSCCAOTDeserializer::invalidateClass(J9VMThread *vmThread, J9Class *ramClass)
+   {
+   assertExclusiveVmAccess(vmThread);
+
+   auto p_it = _classPtrMap.find(ramClass);
+   if (p_it == _classPtrMap.end()) // Not cached or already marked invalid
+      return;
+
+   uintptr_t id = p_it->second;
+   auto i_it = _classIdMap.find(id);
+   TR_ASSERT(i_it != _classIdMap.end(), "Broken two-way map");
+
+   if (i_it->second._ramClass)
+      {
+      // Class ID is valid. Keep the entry, but mark it as unloaded.
+      i_it->second._ramClass = NULL;
+      }
+   else
+      {
+      // Class ID is invalid. Remove the entry so that it can be replaced
+      // if a matching version of the class is ever loaded in the future.
+      _classIdMap.erase(i_it);
```

Comment:
That I can remove, I think. I should just be able to set it to `NULL` unconditionally. You can see that the old `LocalSCC` implementation would also add a `{class, ID}` entry to the `_classPtrMap` when it detected a class mismatch in `cacheRecord` for class serialization records, but that wasn't necessary for me to do, so I only inserted `{ID, NULL}` into the `_classIdMap` and left the pointer map alone.

Reply:
Oh, never mind. I see why it was done that way in the old implementation. If we fail to deserialize a class record then the ID isn't reported to be new, so the server will always send over that class record if a serialized method depends on it. Because of that, in the old way both `{class, ID}` and `{ID, NULL}` would be stored in the maps so that if we got that class record from the server, we would be able to tell from the `{ID, NULL}` entry that we already tried and failed to deserialize that record. Otherwise we'd end up looking up the same mismatching class and failing to validate it.

But, if the class we found during the original deserialization were unloaded, there might be a chance that a new class might be loaded in its place would actually match the class record we'd get from the server. So, if such an unload happens we can safely delete both entries from the map and retry deserialization.

Okay, so I'll change what I've done and have both `{class, ID}` and `{ID, NULL}` stored in the maps if there is a class mismatch. Then the code in `invalidateClass` should make sense as-is.

Reply:
Should be addressed in the latest commit.